From 61491ce54138ed3a23494f1349387d7ac07fcf07 Mon Sep 17 00:00:00 2001
From: Lulu Cheng <chenglulu@loongson.cn>
Date: Wed, 9 Nov 2022 09:47:02 +0800
Subject: [PATCH v1] LoongArch: Complete support for ADDRESS_CONST_INT.

gcc/ChangeLog:

	* config/loongarch/loongarch.cc (loongarch_classify_address):
	Add processing for CONST_INT.
---
 gcc/config/loongarch/loongarch.cc | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/gcc/config/loongarch/loongarch.cc b/gcc/config/loongarch/loongarch.cc
index 7b4d7c1ec51..b266fd18d25 100644
--- a/gcc/config/loongarch/loongarch.cc
+++ b/gcc/config/loongarch/loongarch.cc
@@ -106,7 +106,7 @@ along with GCC; see the file COPYING3.  If not see
        operand is a symbolic address.
 
    ADDRESS_CONST_INT
-       A signed 16-bit constant address.
+       A signed 12-bit constant address.
 
    ADDRESS_SYMBOLIC:
        A constant symbolic address.  */
@@ -2070,6 +2070,12 @@ loongarch_classify_address (struct loongarch_address_info *info, rtx x,
 	      && loongarch_valid_lo_sum_p (info->symbol_type, mode,
 					   info->offset));
 
+    case CONST_INT:
+      /* Small-integer addresses don't occur very often, but they
+	 are legitimate if $r0 is a valid base register.  */
+      info->type = ADDRESS_CONST_INT;
+      return IMM12_OPERAND (INTVAL (x));
+
     default:
       return false;
     }
-- 
2.31.1

